autode/pes/reactive.py

```diff
@@ -0,0 +1,351 @@
+            # Find *thee* TS guess by traversing the minimum energy pathway
+            # from the origin species (reactant state) to a product
+            try:
+                yield next(self._mep_ts_guess(product=product))
+
+            except StopIteration:
+                logger.warning('Found no TS guesses from the minimum energy '
+                               'path')
+
+            return StopIteration
+
+        yielded_p = []
+
+        for idx, point in enumerate(self._sorted_saddle_points()):
+
+            if any(self._distance(p, point) < min_separation for p in
+                   yielded_p):
+                continue
+
+            species = self._species.new_species(name=f'ts_guess{idx}')
+            species.coordinates = self._coordinates[point]
+
+            yielded_p.append(point)
+            yield TSguess.from_species(species)
+
+        return StopIteration
+
+    def _stationary_points(self,
+                           threshold:      float = 0.05,
+                           require_energy: bool = True
+                           ) -> Iterator[Tuple]:
+        """
+        Stationary points on the surface, characterised by a zero gradient
+        vector. On a finite surface the gradient (g) will never truly vanish,
+        so this function will only return those with |g| less than a threshold
+        for which all surrounding points have a larger |g|. Is this a
+        sufficient condition? who knows.
+
+        -----------------------------------------------------------------------
+        Arguments:
+            threshold: Maximum |g| (Ha / Å-1) which to consider a
+                       stationary point.
+
+        Yields:
+            (tuple(int)): Indices of a stationary point
+        """
+        self._set_gradients()
+
+        for point in self._points():
+
+            if require_energy and not self._has_energy(point):
+                continue
+
+            grad = self._gradients[point]
+
+            if np.linalg.norm(grad) > threshold:
+                continue
+
+            if self._is_minimum_in_gradient(point):
+                yield point
+
+        return StopIteration
+
+    def _saddle_points(self,
+                       threshold: float = 0.2
+                       ) -> Iterator[Tuple]:
+        """
+        Find all the saddle points on the surface
+
+        -----------------------------------------------------------------------
+        Arguments:
+            threshold: Threshold on |λ|, below which they are set to zero. Used
+                       to discard small negative eigenvectors which would not
+                       be negative on a finer surface
+
+        Yields:
+            (tuple(int)): Indices of a saddle point
+        """
+        for point in self._stationary_points():
+
+            self._set_hessian(point)
+            eigvals = np.linalg.eigvals(self._hessians[point])
+
+            if sum(lmda < 0 and abs(lmda) > np.max(np.abs(eigvals)) * threshold
```

Comment:
this is super confusing, not really sure what its doing?

Reply:
excellent point - will abstract into a function 👍